src/compiler/scala/tools/nsc/transform/patmat/PatternMatching.scala

```diff
@@ -52,6 +56,30 @@ trait PatternMatching extends Transform
 
   val phaseName: String = "patmat"
 
+  /** Create a new phase which applies transformer */
+  override def newPhase(prev: scala.tools.nsc.Phase): StdPhase = new MyPhase(prev)
+
+  /** The phase defined by this transform */
+  class MyPhase(prev: scala.tools.nsc.Phase) extends StdPhase(prev) {
+    override def run() {
+      echoPhaseSummary(this)
+
+      val threads = scala.util.Properties.envOrElse("PATMAT_THREADS", "4").toInt
+
+      if (threads == 0) {
+        currentRun.units.foreach(applyPhase)
+      } else {
+        implicit val ec = ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(threads))
+        val future = Future.traverse(currentRun.units)(unit => Future(applyPhase(unit)))
+        Await.ready(future, Duration.Inf)
+      }
+    }
+
+    def apply(unit: global.CompilationUnit) {
+      newTransformer(unit).transformUnit(unit)
+    }
+  }
+
   def newTransformer(unit: CompilationUnit): Transformer = new MatchTransformer(unit)
```

Comment:
separate issue - why a new transformer

This may be useful though to hold local state